### DISCUSSION TITLE
"[fx1]Theatre[/fx1]" without any animation

### DISCUSSION THREAD

**Discussion by heartcraftcat** [nndda/godot-theatre#30] [Help] [answered]
Hello nndda. :)
First, I wanted to take a moment to express my sincere gratitude for developing such a high-quality dialogue plugin. It has been a tremendous asset to my project and has significantly enhanced the user experience.

However, I'm encountering a minor issue that I was hoping you could assist me with. In the demo you provided, the "theatre" text featured a stunning blue floating animation effect, which was incredibly impressive. Unfortunately, when I attempt to replicate this in my tests, it only displays as "[fx1]Theatre[/fx1]" without any animation. Could you kindly guide me on how to resolve this issue?

Furthermore, I'm quite intrigued by the range of text animation effects that your plugin supports. Is there a comprehensive list or documentation where I can find all the supported effects and their respective implementation methods? It would be invaluable to me as I explore the full potential of your plugin.

Once again, thank you for your incredible work and for any assistance you can provide in resolving this matter.
shayushenyin@hotmail.com

**Comment by nndda** [comment, marked as answer]
Hi! thank you so much for trying out theatre ^^

These text effects used are actually [Godot Engine's built-in features!](https://docs.godotengine.org/en/stable/classes/class_richtexteffect.html)

As for the text effects `[fx1]`, the effects is defined as a RichTextEffect class in a script. Here's the script used for that effect specifically.

https://github.com/nndda/Theatre/blob/3e7eca8a7f7a1a6c3bbf418475738bb452c33c31/theatre_demo/preview/txt_fx1.gd#L1-L15

And then you can attach it in the `RichTextLabel` or `DialogueLabel`'s `custom_effects` property via the inspector dock:

![image](https://github.com/user-attachments/assets/5703dcb4-039b-4db1-acd5-6ba0d2da30f6)

There are 3 text effects used in the preview dialogue scene. All of them are located at:
```
theatre_demo/preview/txt_fx1.gd
theatre_demo/preview/txt_fx2.gd
theatre_demo/preview/txt_fx3.gd
```

Here are some resources about RichTextEffect that you might find helpful:
- https://docs.godotengine.org/en/stable/classes/class_richtexteffect.html
- https://www.youtube.com/watch?v=8CZfqdUd3bM
- https://github.com/Goldenlion5648/CustomRichTextEffectsDemo
- https://github.com/chairfull/GodotRichTextLabel2

Hope this helps! ^^

**Reply by heartcraftcat** [reply]
REALLY Thank you for your prompt and kind response.

By the way， I was wondering if it would be possible to add a new feature to Theatre: the ability to read a CSV dialogue file. for example, the CSV file could have the following columns:

Column 1: Sequence Number - Representing the order or sequence of the dialogue.
Column 2: Character Name - Indicating the name of the character speaking.
Column 3: Dialogue Content - Containing the actual text of what the character is saying.
Column 4: Audio File URL - A URL or path to where the voiceover or audio recording of the dialogue is stored.
Column 5: Function Call (Optional) - This column could potentially hold references to functions that can be called when the dialogue is triggered, allowing for additional interactivity or dynamic behavior based on the dialogue.

Having such a feature would greatly enhance the flexibility and usability of the system, allowing for easy integration of pre-recorded dialogues and potentially triggering custom functions or events based on the content of the dialogue. 

Thanks！

**Reply by nndda** [reply]
Glad to help ^_^ @shayushenyin 

> I was wondering if it would be possible to add a new feature to Theatre: the ability to read a CSV dialogue file.

I've actually considered using CSV file format (although for different reason and approach) #15

Honestly, CSV seems easier to process and work with. I'll reconsider adding it as an alternative format for Dialogue resources. But for now, I'll stick with the current `.dlg` format.

> Having such a feature would greatly enhance the flexibility and usability of the system, allowing for easy integration of pre-recorded dialogues and potentially triggering custom functions or events based on the content of the dialogue.

I want to leave specifics implementations up to users (audio speech, character portraits, etc.). But I can see that it might be not enough with just the current function calls features.

There's also `progressed_at`, `skipped_at`, and `cancelled_at` signals in `Stage`, that you can use to do things based off the Dialogue data:
```gdscript
func _on_stage_progressed_at(line: int, data: Dictionary):
    if data.actor == "Dia":
        # ...

    if data.line == "Hello, world!":
        # ...
```
But it is still very much undocumented. And the `data` of the Dialogue line can be very unintuitive.

So I plan to add things like Dialogue line attributes/properties, and user-defined custom Dialogue tags. Maybe something like:
```
{player}:
    audio_file="res://player_speech.wav"
    portrait="res://player_smile.png"

    "..."
```
